plugin/rewrite/name.go

```diff
@@ -1,29 +1,116 @@
+    PartialMatch = "partial"
+    // Execute rewrite on when the name in the question section of a request matches a regular expession
+    RegexMatch = "regex"
+)
+
+// Rewrite rewrites the the current request based upon exact match of the name
+// in the question section of the request
 func (rule *nameRule) Rewrite(w dns.ResponseWriter, r *dns.Msg) Result {
-	if rule.From == r.Question[0].Name {
-		r.Question[0].Name = rule.To
-		return RewriteDone
-	}
-	return RewriteIgnored
+    if rule.From == r.Question[0].Name {
+        r.Question[0].Name = rule.To
+        return RewriteDone
+    }
+    return RewriteIgnored
+}
+
+// Rewrite rewrites the the current request based upon partial match of the
+// name in the question section of the request
+func (rule *partialNameRule) Rewrite(w dns.ResponseWriter, r *dns.Msg) Result {
+    if strings.Contains(r.Question[0].Name, rule.Substring) {
+        r.Question[0].Name = strings.Replace(r.Question[0].Name, rule.Substring, rule.Replacement, -1)
+        return RewriteDone
+    }
+    return RewriteIgnored
+}
+
+// Rewrite rewrites the the current request when the name in the question
+// section of the request matches a regular expession
+func (rule *regexNameRule) Rewrite(w dns.ResponseWriter, r *dns.Msg) Result {
+    regexGroups := rule.Pattern.FindStringSubmatch(r.Question[0].Name)
+    if len(regexGroups) == 0 {
+        return RewriteIgnored
+    }
+    s := rule.Replacement
+    for groupIndex, groupValue := range regexGroups {
+        groupIndexStr := "{" + strconv.Itoa(groupIndex) + "}"
+        if strings.Contains(s, groupIndexStr) {
+            s = strings.Replace(s, groupIndexStr, groupValue, -1)
+        }
+    }
+    r.Question[0].Name = s
+    return RewriteDone
+}
+
+// newNameRule creates a name matching rule based on exact, partial, or regex match
+func newNameRule(nextAction string, args ...string) (Rule, error) {
+    if len(args) < 2 {
+        return nil, fmt.Errorf("too few arguments for a name rule")
+    }
+    if len(args) > 3 {
+        return nil, fmt.Errorf("exceeded the number of arguments for a name rule")
+    }
+    if len(args) == 3 {
+        switch strings.ToLower(args[0]) {
+        case ExactMatch:
+            return &nameRule{nextAction, plugin.Name(args[1]).Normalize(), plugin.Name(args[2]).Normalize()}, nil
+        case PartialMatch:
+            return &partialNameRule{nextAction, plugin.Name(args[1]).Normalize(), plugin.Name(args[2]).Normalize()}, nil
+        case RegexMatch:
+            regexPattern, err := regexp.Compile(args[1])
+            if err != nil {
+                return nil, fmt.Errorf("Invalid regex pattern in a name rule: %s", args[1])
+            }
+            return &regexNameRule{nextAction, regexPattern, plugin.Name(args[2]).Normalize()}, nil
+        default:
+            return nil, fmt.Errorf("A name rule supports only exact, partial, and regex name matching")
```

Comment:
This is not backwards compatible.

`rewrite name foo.com bar.com`

will now fail to parse. I think you need a case with 2 args.